src/main/java/server/ChatServer.java

```diff
@@ -8,118 +8,78 @@
-            s = ss.accept();
-
-            System.out.println("New client request received : " + s);
-
-            // obtain input and output streams
-            DataInputStream dis = new DataInputStream(s.getInputStream());
-            PrintWriter dos = new PrintWriter(s.getOutputStream(),true);
-            StringTokenizer st = new StringTokenizer(dis.readLine(), "#");
-            String cmd = st.nextToken();
-
-            try {
-
-
-                if (cmd.equals("CONNECT")) {
-                    String username = st.nextToken();
-                    int i = 0;
-
-
-                    for (String users : userList) {
-
-
-                        if (username.equals(users)) {
-
-                            System.out.println("Creating a new handler for this client...");
-
-                            // Create a new handler object for handling this request.
-                            ClientHandler mtch = new ClientHandler(s, username, dis, dos);
-
-                            // Create a new Thread with this object.
-                            Thread t = new Thread(mtch);
-
-                            System.out.println("Adding this client to active client list");
-
-                            // add this client to active clients list
-                            ar.add(mtch);
-
-                            // start the thread.
-                            t.start();
-
-                            onlineMessage();
-
-                            break;
-
-                        }
-                        i++;
-
-                        if (i == userList.size()) {
-                            dos.println("CLOSE#2");
-                            s.close();
-                            break;
-                        }
-
-
-                    }
+    /**
+     * sendMessage attempts to send a message to the given destination.
+     */
+    public static void sendMessage(ClientHandler source, String destination, String message) {
+        StringBuilder builder = new StringBuilder();
+        builder.append("MESSAGE#");
+        if (destination == "*") {
+            builder.append("*");
+        } else {
+            builder.append(source.getName());
+        }
 
+        builder.append("#");
+        builder.append(message);
 
-                }
 
+        String msg = builder.toString();
+        for (ClientHandler mc : ChatServer.ar) {
+            if (!mc.nameMatches(pattern)) {
+                continue;
+            }
 
-                if (!cmd.equals("CONNECT")) {
-                    dos.println("CLOSE#1");
-                    s.close();
+            mc.sendMessage(msg);
+        }
+    }
 
-                }
+    public static void main(String[] args) throws IOException {
+        UserService userService = new UserService();
+        ServerSocket ss = new ServerSocket(8000);
+        System.out.println("Server started");
 
-            } catch (Exception e) {
+        while (true) {
```

Comment:
This is heavily simplified code of the previous loop. Most of the logic is handled in `tryAccept`. In reality you would probably use exceptions to indicate failures in accepting in Java rather than checking if the result is null (or, betteryet, you'd return a ClientHandler which did nothing, but in order to do that you'd need more complicated changes that go beyond the scope of this PR)